internal/codespaces/portforwarder/port_forwarder.go

```diff
@@ -128,8 +164,57 @@ func (fwd *PortForwarder) ForwardPort(ctx context.Context, remotePort uint16, li
 	}
 }
 
+// connectToForwardedPort connects to the forwarded port via a local TCP port or a given ReadWriteCloser.
+// Optionally, it detects traffic over the connection and sends activity signals to the server to keep the codespace from shutting down.
+func (fwd *CodespacesPortForwarder) connectToForwardedPort(ctx context.Context, portNumber uint16, keepAlive bool, listener *net.TCPListener, conn io.ReadWriteCloser) (err error) {
+	errc := make(chan error, 1)
+	sendError := func(err error) {
+		// Use non-blocking send, to avoid goroutines getting
+		// stuck in case of concurrent or sequential errors.
+		select {
+		case errc <- err:
+		default:
+		}
+	}
+	go func() {
+		for {
+			// Make a copy of the connection so we can accept new connections before this one closes
+			connCopy := conn
+
+			// If a listener is provided, accept a connection from it
+			if listener != nil {
+				connCopy, err = listener.AcceptTCP()
+				if err != nil {
+					sendError(err)
+					return
+				}
+			}
+
+			// Create a traffic monitor to keep the session alive
+			if keepAlive {
+				connCopy = newTrafficMonitor(connCopy, fwd)
+			}
+
+			// Connect to the forwarded port in a goroutine so we can accept new connections
```

Comment:
If we are using the ReadWriteCloser and not the TCPListener, is it valid to loop here or should we exit after the first connection? Without the listener won't it just spam connections to the ReadWriteCloser?

Reply:
Good callout, that definitely seems to be the case and was bogging down the performance of the command when using the `--stdio` flag.

Reply:
Fun fact, that's actually an existing problem in the current release and it is now fixed!